File: mwe_script.r
library(reticulate)
library(igraph)
library(tidyverse)


use_virtualenv("./venv")

moo <- import("moo")
moo_datagen <- import("moo.data_generation")
contestant <- import("moo.contestant")
moo_communities <- import("moo.communities")
multicriteria <- import("moo.multicriteria")
pd <- import("pandas")


expconfig <- moo_datagen$ExpConfig(
  L=list(500L,500L,500L,500L,500L), U=list(500L,500L,500L,500L,500L),
  NumEdges=7500L, 
  BC=0.1, NumGraphs=100L, # Was 30L
  shuffle=TRUE, 
  seed=1234L  
  
)

expconfig <- moo_datagen$ExpConfig(
  L=list(150L,150L), U=list(150L,150L),
  NumEdges=200L, 
  BC=0.1, NumGraphs=1L, # Was 30L
  shuffle=TRUE, 
  seed=24L  
  
)

algos = list(
  contestant$ComDetMultiLevel(), # Multi-Level approach
  #contestant$ComDetEdgeBetweenness(), # EdgeBetweenness approach
  #contestant$ComDetWalkTrap(), # WalkTrap approach
  #contestant$ComDetFastGreedy(), # FastGreedy approach
  contestant$ComDetBRIM(max_num_clusters=1L)#, # Brim
  #multicriteria$ComDetMultiCriteria(name = "3d",
  #                                  params = list(
  #                                    'mode'= '3d', # '2d' for 2d approach
  #                                    'popsize'= 50L,
  #                                    'termination'= py_none(), # By default it runs for 1000 generations (or pass a pymoo termination instance)
  #                                    'save_history'= FALSE, # set to True for later hypervolume calculations
  #                                    'seed'= py_none() # For reproducibility
  #                                  ))
)

## The error from Luis is as below.
moo_multicriteria <- import("moo.multicriteria")
pymoo_termination <- import("pymoo.util.termination.f_tol")
algos_2d_newmut_enh = list(moo_multicriteria$ComDetMultiCriteria(name="2d_newmut_enh",
			params = list(
			'mode'='2d',
			'popsize'=50L,
			'termination'=pymoo_termination$MultiObjectiveSpaceToleranceTermination(tol=0.0001,
									n_last=100L,
									nth_gen=75L,
									n_max_gen=5000L,
									n_max_evals=py_none()),
			'save_history'=FALSE,
			'seed'=py_none(),
			'initialization'='original',
			'mutation'='enhanced',
			'enhance'=0.2,
			'pair_crossover'=1,
			'gene_crossover'=0.1
)) )


expgen = moo_datagen$DataGenerator(expconfig=expconfig) # Pass defined parameters

datagen <- expgen$generate_data()

results <- moo_communities$run_serial_communities(datagen, algos_2d_newmut_enh)
results <- moo_communities$run_parallel_communities(datagen, algos)

# I've been unable to get parallel processing working on Windows.  It's fine
# on Windows Subsystem for Linux (WSL).  I strongly suspect it'll be fine on 
# "proper" Linux and macos.  
# Have tried using fix at https://github.com/rstudio/reticulate/issues/517
# results <- moo_communities$run_parallel_communities(datagen, algos, n_jobs=2)

results_df = pd$DataFrame(results)

best_solutions <- contestant$get_best_community_solutions(results_df)


best_solutions%>% as_tibble() %>% 
  ggplot(aes(x = name, y = adj_rand_index)) + 
  geom_boxplot()

## Get the convergence plot for the first algo.
alg = algos[1]
vol_dat = alg$compute_hypervolume()

File: mwe_script.py
## This script is a minimum working example of the data process.

import igraph
import pandas as pd
from moo.data_generation import ExpConfig, DataGenerator
from moo.contestant import get_best_community_solutions, draw_best_community_solutions
import moo.contestant as contestant
from moo.multicriteria import ComDetMultiCriteria
import matplotlib.pyplot as plt
import sknetwork
import code
import time
from copy import deepcopy
import pymoo.util.termination.f_tol

start = time.time()

## Run the data loading.
expconfig = ExpConfig(
    L=[50,50], U=[50,50], NumEdges=200, BC=0.1, NumGraphs=1,
    shuffle=True, filename='test_graphs_', seed=24#42
)

print(expconfig) # Print parameters, or access individually, e.g., expconfig.NumEdges

print('Config defined in %f s' % (time.time()-start))
start = time.time()

# Generate data following the defined experiment confguration
expgen = DataGenerator(expconfig=expconfig) # Pass defined parameters
print(expgen)
datagen = expgen.generate_data() # datagen is an iterator

print('Data generator constructed in %f s' % (time.time()-start))
start = time.time()

## Define the algorithms.
algos = [
    contestant.ComDetMultiLevel(), # Multi-Level approach
    contestant.ComDetEdgeBetweenness(), # EdgeBetweenness approach
    contestant.ComDetWalkTrap(), # WalkTrap approach
    contestant.ComDetFastGreedy(), # FastGreedy approach
    contestant.ComDetBRIM(), # BRIM approach
    contestant.ComDetBiLouvain(), # sknetwork bilovain
    #contestant.ComDetariadne(), # pymocd ariadne
    #contestant.ComDethpmocd(), # pymocd hpmocd
    #contestant.ComDetmocd_q(), # pymocd mocd_q
    #contestant.ComDetmocd_d(), # pymocd mocd_d
    #contestant.ComDetmoga_net(), # pymocd moga_net (Pizzuti)
    #contestant.ComDetccm(), # pymocd ccm
    #contestant.ComDetkrm(), # pymocd krm
    #contestant.ComDetmmcomo(), # pymocd mmcomo
    #ComDetMultiCriteria(  # 3D MO approach
    #name='3d',
    #params = {
    #    'mode': '3d', # '2d' for 2d approach
    #    'popsize': 50,
    #    'termination': None, # By default it runs for 1000 generations (or pass a pymoo termination instance)
    #    'save_history': False, # set to True for later hypervolume calculations
    #    'seed': None, # For reproducibility
    #    'initialization': 'original',
    #    'mutation': '',#'enhanced',
    #    'enhance':0,
    #}
#),
#    ComDetMultiCriteria(  # 2D MO approach
 #   name='2d',
  #  params = {
   #     'mode': '2d', # '2d' for 2d approach
    #    'popsize': 50,
     #   'termination': pymoo.util.termination.f_tol.MultiObjectiveSpaceToleranceTermination(tol=0.0001,n_last=100,nth_gen=75,n_max_gen=5000,n_max_evals=None),#None, # By default it runs for 1000 generations (or pass a pymoo termination instance)
      #  'save_history': False, # set to True for later hypervolume calculations
       # 'seed': None, # For reproducibility
        #'initialization': 'original',
        #'mutation': 'enhanced',#'enhanced',
        #'enhance':0.2,
        #'pair_crossover':1.0,
        #'gene_crossover':0.1
    #}
#)
]

print('Algos defined in %f s' % (time.time()-start))
start = time.time()

## Detect some communities.
results = [] # Holds results of contestants
for g_idx, graph in enumerate(datagen):
    print(f'Processing Graph {g_idx+1}')
    
    ## Generate badj to pass and save computation time.
    #code.interact(local=locals())
    for algo in algos:
        print(f'\tUsing algoithm {algo.name_}')
        result = algo.detect_communities(graph=deepcopy(graph)).get_results()
        # Result is a list of dictionaries, each dictionary stores the metrics of one iteration (see code for details)
        for r in result: # Appending graph index to results, for debugging purposes
            r['graph_idx'] = g_idx + 1
        results.extend(result)
        print('Results collected in %f s' % (time.time()-start))
        start = time.time()

df_contestants = pd.DataFrame(results)
best_solutions = get_best_community_solutions(df_contestants).reset_index(drop=True)

fig, ax = plt.subplots(figsize=(15,8))
ax, stats = draw_best_community_solutions(best_solutions,ax)

code.interact(local=locals())